lib/cc/analyzer/engine.rb

```diff
@@ -35,32 +34,30 @@ def run(stdout_io, stderr_io = StringIO.new)
           end
         end
 
-        Thread.new do
-          sleep TIMEOUT
-
-          if engine_running
-            Thread.current.abort_on_exception = true
-            run_command("docker kill #{container_name}")
-
-            Analyzer.statsd.increment("cli.engines.result.error")
-            Analyzer.statsd.increment("cli.engines.result.error.timeout")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error")
-            Analyzer.statsd.increment("cli.engines.names.#{name}.result.error.timeout")
-            raise EngineTimeout, "engine #{name} ran past #{TIMEOUT} seconds and was killed"
+        begin
+          status = nil
+          Timeout::timeout(TIMEOUT) do
+            pid, status = Process.waitpid2(pid)
           end
+        rescue Timeout::Error
+          run_command("docker kill #{container_name} || true")
+          timed_out = true
```

Comment:
Can we get rid of `timed_out` and do whatever happens on `timed_out = true` in the `rescue`? I think the only change in behavior is that `"cli.engines.finished"` wouldn't be ticked in the case of a timeout. If that's important, I'd just duplicate the `increment` call: once in the happy path and again in the timeout rescue.

Also, the `begin`/`rescue` is telling me to extract a `#wait_for_container` method.